File: src/main/java/com/somefrills/features/misc/glowmob/chestui/ChestUI.java
package com.somefrills.features.misc.glowmob.chestui;

import com.somefrills.misc.Utils;
import net.minecraft.client.gui.screen.ingame.GenericContainerScreen;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.screen.GenericContainerScreenHandler;
import net.minecraft.screen.slot.Slot;
import net.minecraft.screen.slot.SlotActionType;
import net.minecraft.text.Text;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static com.somefrills.Main.mc;

public abstract class ChestUI extends GenericContainerScreen {
    protected static final int INV_SIZE = 9 * 6;
    private static final long CLICK_COOLDOWN_MS = 50L;

    protected final List<UIAddon> addons = new ArrayList<>();
    protected final List<ItemStack> allItems = new ArrayList<>();
    protected final ChestUI previousScreen;
    protected long lastClickTimestamp = 0L;

    public ChestUI(String title) { this(title, null); }

    public ChestUI(String title, ChestUI previousScreen) {
        super(getHandler(INV_SIZE), Objects.requireNonNull(mc.player).getInventory(), Text.of(title));
        this.previousScreen = previousScreen;
        addAddon(new CloseAddon());
    }

    public void addAddon(UIAddon addon) {
        this.addons.add(addon);
    }

    public void clearAddons() {
        this.addons.clear();
    }

    public static GenericContainerScreenHandler getHandler(int invSize) {
        int syncId = 0;
        Inventory inventory = new UIInventory(invSize);
        return GenericContainerScreenHandler.createGeneric9x6(syncId, mc.player.getInventory(), inventory);
    }

    protected Inventory getInventory() {
        return handler.getInventory();
    }

    public final void rebuild() {
        Inventory inv = getInventory();
        inv.clear();
        fillBorder(inv);

        // 1. Let the subclass define what items exist
        allItems.clear();
        build();

        // 2. Run addons to filter or paginate the list
        List<ItemStack> displayList = new ArrayList<>(allItems);
        for (UIAddon addon : addons) {
            addon.processItems(this, displayList);
        }

        // 3. Render the processed list into the UI
        renderList(inv, displayList);

        // 4. Let addons draw their navigation buttons
        for (UIAddon addon : addons) {
            addon.drawDecoration(this, inv);
        }
    }

    protected abstract void build();

    private void renderList(Inventory inv, List<ItemStack> items) {
        int slotPtr = 0;
        for (ItemStack stack : items) {
            while (slotPtr < inv.size() && isBorderSlot(slotPtr)) {
                slotPtr++;
            }
            if (slotPtr >= inv.size()) break;
            inv.setStack(slotPtr++, stack);
        }
    }

    protected void fillBorder(Inventory inventory) {
        for (int i = 0; i < inventory.size(); i++) {
            if (isBorderSlot(i)) {
                inventory.setStack(i, new ItemStack(Items.GRAY_STAINED_GLASS_PANE));
            }
        }
    }

    protected boolean isBorderSlot(int slotIndex) {
        return slotIndex < 9 || slotIndex >= INV_SIZE - 9 || slotIndex % 9 == 0 || slotIndex % 9 == 8;
    }

    protected void onItemClick(ItemStack stack, int button) {}

    @Override
    public void onMouseClick(Slot slot, int slotId, int button, SlotActionType actionType) {
        if (slot == null || !slot.hasStack()) return;

        long now = System.currentTimeMillis();
        if (now - this.lastClickTimestamp < CLICK_COOLDOWN_MS) return;
        this.lastClickTimestamp = now;

        ItemStack stack = slot.getStack();
        String name = Utils.getPlainCustomName(stack);
        if (name == null) return;

        // Addon Interception
        for (UIAddon addon : addons) {
            if (addon.onClick(this, stack, name, button)) return;
        }

        // Hardcoded Close Logic
        if (name.equals("Close")) {
            this.close();
            return;
        }

        onItemClick(stack, button);
    }

    protected void addItem(ItemStack stack) {
        allItems.add(stack);
    }

    protected void onReturn() {
        rebuild();
    }

    @Override
    public void close() {
        if (this.client.player != null) this.client.player.closeHandledScreen();
        if (this.previousScreen != null) {
            this.previousScreen.onReturn();
            Utils.setScreen(this.previousScreen);
        }
    }

}

File: src/main/kotlin/com/somefrills/features/misc/glowmob/chestui/ChestUI.kt
package com.somefrills.features.misc.glowmob.chestui

import com.somefrills.Main.mc
import com.somefrills.utils.GuiUtils
import com.somefrills.utils.plainCustomName
import io.github.notenoughupdates.moulconfig.common.ClickType
import net.minecraft.client.gui.screens.inventory.ContainerScreen
import net.minecraft.network.chat.Component
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.inventory.ChestMenu
import net.minecraft.world.inventory.ContainerInput
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import kotlin.collections.ArrayList
import kotlin.collections.MutableList

abstract class ChestUI @JvmOverloads constructor(
    title: String,
    @JvmField val previousScreen: ChestUI? = null
) : ContainerScreen(
    getHandler(INV_SIZE),
    mc.player!!.inventory,
    Component.literal(title)
) {

    protected val addons: MutableList<UIAddon> = ArrayList()

    @JvmField
    protected val allItems: MutableList<ItemStack> = ArrayList()

    protected var lastClickTimestamp: Long = 0L


    init {
        addAddon(CloseAddon())
    }


    fun addAddon(addon: UIAddon?) {
        addons.add(addon!!)
    }

    fun clearAddons() {
        addons.clear()
    }


    protected val inventory: Inventory
        get() = this.minecraft!!.player!!.inventory


    fun rebuild() {
        val inv = this.menu.container

        inv.clearContent()
        fillBorder(inv)

        allItems.clear()
        build()

        val displayList = ArrayList(allItems)

        for (addon in addons) {
            addon.processItems(this, displayList)
        }

        renderList(inv, displayList)

        for (addon in addons) {
            addon.drawDecoration(this, inv)
        }
    }


    protected abstract fun build()


    private fun renderList(inv: net.minecraft.world.Container, items: MutableList<ItemStack>) {
        var slotPtr = 0

        for (stack in items) {
            while (slotPtr < inv.containerSize && isBorderSlot(slotPtr)) {
                slotPtr++
            }

            if (slotPtr >= inv.containerSize) {
                break
            }

            inv.setItem(slotPtr++, stack)
        }
    }


    protected fun fillBorder(inventory: net.minecraft.world.Container) {
        for (i in 0 until inventory.containerSize) {
            if (isBorderSlot(i)) {
                inventory.setItem(
                    i,
                    ItemStack(Items.GRAY_STAINED_GLASS_PANE)
                )
            }
        }
    }


    protected fun isBorderSlot(slotIndex: Int): Boolean {
        return slotIndex < 9 ||
                slotIndex >= INV_SIZE - 9 ||
                slotIndex % 9 == 0 ||
                slotIndex % 9 == 8
    }


    protected open fun onItemClick(stack: ItemStack?, button: Int) {
    }

    override fun slotClicked(
        slot: Slot, slotId: Int, buttonNum: Int, containerInput: ContainerInput
    ) {
        if (!slot.hasItem()) return

        val now = System.currentTimeMillis()

        if (now - lastClickTimestamp < CLICK_COOLDOWN_MS) {
            return
        }

        lastClickTimestamp = now

        val stack = slot.item
        val name = stack.plainCustomName

        for (addon in addons) {
            if (addon.onClick(this, stack, name, buttonNum)) {
                return
            }
        }

        if (name == "Close") {
            super.onClose()
            return
        }

        onItemClick(stack, buttonNum)
    }


    protected fun addItem(stack: ItemStack) {
        allItems.add(stack)
    }


    protected open fun onReturn() {
        rebuild()
    }


    override fun onClose() {
        if (minecraft?.player != null) {
            minecraft!!.player!!.closeContainer()
        }

        previousScreen?.let {
            it.onReturn()
            GuiUtils.setScreen(it)
        }
    }


    companion object {

        const val INV_SIZE = 9 * 6
        private const val CLICK_COOLDOWN_MS = 50L


        fun getHandler(invSize: Int): ChestMenu {
            val syncId = 0

            val container = UIInventory(invSize)

            return ChestMenu.sixRows(
                syncId,
                mc.player!!.inventory,
                container
            )
        }
    }
}